### PR TITLE
Add a 'RunOnRepl' module

### PR DESCRIPTION
**Pull request by dnadales**
... which allows to (re-)test a counterexample found by the `ChainDB.StateMachine` tests by copying and pasting it in the GHC repl.

Part of #4183 

# Checklist

- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+--
+-- > quickCheckeCmdsLockStep someClockSkew someChunkInfo counterexample
+--
+-- Where 'someClockSkew' and 'someChunkInfo' are the ones given by a the
+-- counterexample found by quickcheck-statemachine.
+module Test.Ouroboros.Storage.ChainDB.StateMachine.RunOnRepl (
+    -- * Running the counterexamples
+    quickCheckeCmdsLockStep
+    -- * Patterns needed to disambiguate the 'At' and 'Command' symbols printed
+    -- * by the 'ChainDB.StateMachine' tests.
```
Comment (abailly-iohk) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs:41]: typo: Remove leading `*` as I don't think you want another list item here

**Review by abailly-iohk — APPROVED** [review]

File: ouroboros-consensus-test/ouroboros-consensus-test.cabal [hunk 2]
```diff
@@ -228,6 +228,7 @@ test-suite test-storage
                        Test.Ouroboros.Storage.ChainDB.Model.Test
                        Test.Ouroboros.Storage.ChainDB.Paths
                        Test.Ouroboros.Storage.ChainDB.StateMachine
+                       Test.Ouroboros.Storage.ChainDB.StateMachine.RunOnRepl
```
Comment (bartfrenk) [inline ouroboros-consensus-test/ouroboros-consensus-test.cabal:231]: nitpick: Why not just Test.Ouroboros.Storage.ChainDB.StateMachine.REPL

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+pattern Command ::
+     t1 blk1 (IterRef blk1 m1 Symbolic) (FollowerRef blk1 m1 Symbolic)
+  -> t2 blk2 (IterRef blk2 m2 Symbolic) (FollowerRef blk2 m2 Symbolic)
+  -> [Var]
+  -> StateMachine.Types.Command (StateMachine.At t1 blk1 m1) (StateMachine.At t2 blk2 m2)
+pattern Command cmd rsp xs =
+  StateMachine.Types.Command (StateMachine.At cmd) (StateMachine.At rsp) xs
+
+quickCheckeCmdsLockStep ::
```
Comment (bartfrenk) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs:119]: Typo?

```suggestion
quickCheckCmdsLockStep ::
```

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+     MaxClockSkew
+  -> SmallChunkInfo
+  -> Commands (StateMachine.At Cmd TestBlock IO) (StateMachine.At Resp TestBlock IO)
+  -> IO ()
+quickCheckeCmdsLockStep maxClockSkew chunkInfo cmds =
```
Comment (bartfrenk) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs:124]:
```suggestion
quickCheckCmdsLockStep maxClockSkew chunkInfo cmds =
```

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs (hunk 1 above)
Comment (dnadales) [reply ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs:41]: Ups, `M-q` :grin: Thanks.

File: ouroboros-consensus-test/ouroboros-consensus-test.cabal (hunk 2 above)
Comment (dnadales) [reply ouroboros-consensus-test/ouroboros-consensus-test.cabal:231]: My thought: it's not a REPL, but rather imports to allow the state machine to be run on the REPL. What do you think?

File: ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+--
+-- Note the use of parentheses to prevent GHCi from ending the multiline input
+-- prematurely (the copied counterexample in this case).
+--
+-- Then, the model and system under tests can be tested for lockstep agreement
+-- by running:
+--
+-- > quickCheckCmdsLockStep someClockSkew someChunkInfo counterexample
+--
+-- Where 'someClockSkew' and 'someChunkInfo' are the ones given by a the
```
Comment (Jasagredo) [inline ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine/RunOnRepl.hs:35]:
```suggestion
-- Where 'someClockSkew' and 'someChunkInfo' are the ones given by the
```

**Review by Jasagredo — APPROVED** [review]
Perhaps we could name this module `.Utils.RunOnRepl` to convey that it is not a test in itself but utilities for the tests?

**Comment by dnadales** [comment]
> Perhaps we could name this module `.Utils.RunOnRepl` to convey that it is not a test in itself but utilities for the tests?

Makes sense. I can rename the module.

**Comment by dnadales** [comment]
bors retry

**Pull request merged**